### PR TITLE
Attempting to fix 738

### LINKED ISSUE #738: SiriusXM/Chromecast Casting Issue
There is an issue regarding Chromecast integration in Home Assistant and the SiriusXM app - if the cast integration is enabled, casting SiriusXM to a Chromecast device will fail with an error code 1001. Disabling cast integration in HASS fixes the problem.

Original discussion here - also confirmed another add-on that uses pychromecast (Music Assistant 2.0 Beta) results in the same behavior:

https://github.com/home-assistant/core/issues/93360

### PR DESCRIPTION
**Pull request by geoffoxholm**
I'm trying to fix #738.

It looks like some apps (maybe just "Audible" at the moment) have a very rigid interpretation of the connection handshake, and abandon it if messages come out-of-order. 

I'd love any feedback to help move this forward. I can step into the debugger, do whatever is needed to figure out how to fix the issue, though I'm new to the `pychromecast` repo.

### CODE REVIEW

**Comment by geoffoxholm** [comment]
I would appreciate some feedback on this. Is there a better approach I could try? Any tips would be appreciated, I'll give it a serious try.

File: pyproject.toml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [build-system]
-requires = ["setuptools~=65.6", "wheel~=0.37.1"]
+requires = ["setuptools~=69.2.0", "wheel~=0.43.0"]
```
Comment (emontnemery) [inline pyproject.toml:2]: This seems unrelated?

File: pychromecast/socket_client.py [hunk 2]
```diff
@@ -489,6 +489,11 @@ def new_cast_status(self, status: CastStatus) -> None:
         self.destination_id = status.transport_id
         self.session_id = status.session_id
 
+        # When a new cast status is received, give Audible time to load
+        if status.app_id == "25456794":
+            self.logger.debug("Detected Audible connection. Sleeping for 1s")
+            time.sleep(1)
```
Comment (emontnemery) [inline pychromecast/socket_client.py:495]: This is not really OK way to fix a race. Is there another message we can wait for instead?

File: pyproject.toml (hunk 1 above)
Comment (geoffoxholm) [reply pyproject.toml:2]: definitely unrelated. i can revert the change.

File: pychromecast/socket_client.py (hunk 2 above)
Comment (geoffoxholm) [reply pychromecast/socket_client.py:495]: i'm not sure. any ideas you have, or example code you can point me to? i'll try whatever i can.

File: pyproject.toml (hunk 1 above)
Comment (emontnemery) [reply pyproject.toml:2]: Please do

File: pychromecast/socket_client.py (hunk 2 above)
Comment (emontnemery) [reply pychromecast/socket_client.py:495]: Issue https://github.com/home-assistant-libs/pychromecast/issues/738 talks about the Sirius XM app misbehaving, but your comment in the PR talks about Audible. Could you clarify the relationship between the two apps? Or are they the same?

Also, after going through the comments in https://github.com/home-assistant-libs/pychromecast/issues/738, it's not fully clear to me what pychromecast does which offends the app so much that it terminates, only that adding a delay fixes the problem.

Have you been able to determine *how* the delay fixes it?

If you run pychromecast with debug enabled, it will dump out all messages sent to and received from the pychromecast, have you done that?

Also, I'm not sure how you've been debugging. But if you want a simple isolated test case, maybe this example script is enough to trigger (and debug) the problem:
```
python3 -m examples.simple_listener_example --cast 'Living room' --show-debug
```

Comment (emontnemery) [reply pychromecast/socket_client.py:495]: Ah, there's already a comment in #738 about using the simple listener example to trigger the problem: https://github.com/home-assistant-libs/pychromecast/issues/738#issuecomment-1650581622

If you could do that, both with and without your workaround, that would be really useful.

Comment (geoffoxholm) [reply pychromecast/socket_client.py:495]:
```sh
git switch master
python3 -m examples.simple_listener_example --cast 'Garage speakers' --show-debug > workspace/master.std.txt 2> workspace/master.err.txt
```
[master.err.txt](https://github.com/user-attachments/files/16825091/master.err.txt)
[master.std.txt](https://github.com/user-attachments/files/16825092/master.std.txt)


```sh
git switch geoffoxholm:users/geoff/audible_debug
python3 -m examples.simple_listener_example --cast 'Garage speakers' --show-debug > workspace/patch.std.txt 2> workspace/patch.err.txt
```
[patch.err.txt](https://github.com/user-attachments/files/16825094/patch.err.txt)
[patch.std.txt](https://github.com/user-attachments/files/16825095/patch.std.txt)

Comment (emontnemery) [reply pychromecast/socket_client.py:495]: There's a jump in `requestId` in messages sent by pychromecast to the device which I can't explain, which is really odd because we only increment the `requestId` when calculating it here: https://github.com/home-assistant-libs/pychromecast/blob/449c0b52261804a2a1077700d6653d69cf902691/pychromecast/socket_client.py#L855-L856

Have you removed any lines from the logs?

Comment (emontnemery) [reply pychromecast/socket_client.py:495]: To be sure, could you change here to log also the pings, both on master and your patch branch, and collect new logs: https://github.com/home-assistant-libs/pychromecast/blob/449c0b52261804a2a1077700d6653d69cf902691/pychromecast/socket_client.py#L873-L882

https://github.com/home-assistant-libs/pychromecast/blob/449c0b52261804a2a1077700d6653d69cf902691/pychromecast/socket_client.py#L689-L697

Comment (geoffoxholm) [reply pychromecast/socket_client.py:495]:
```sh
git switch master
git pull
# make edits
git diff > workspace/master.diff.txt
python3 -m examples.simple_listener_example --cast 'Garage speakers' --show-debug > workspace/master.std.txt 2> workspace/master.err.txt
```
[master.diff.txt](https://github.com/user-attachments/files/16875720/master.diff.txt)
[master.err.txt](https://github.com/user-attachments/files/16875724/master.err.txt)
[master.std.txt](https://github.com/user-attachments/files/16875726/master.std.txt)

```sh
git stash # stash edits
git switch users/geoff/audible_debug 
git stash pop # pop edits
git diff master > workspace/branch.diff.txt
python3 -m examples.simple_listener_example --cast 'Garage speakers' --show-debug > workspace/branch.std.txt 2> workspace/branch.err.txt
```
[branch.diff.txt](https://github.com/user-attachments/files/16875735/branch.diff.txt)
[branch.err.txt](https://github.com/user-attachments/files/16875737/branch.err.txt)
[branch.std.txt](https://github.com/user-attachments/files/16875739/branch.std.txt)

File: pyproject.toml (hunk 1 above)
Comment (geoffoxholm) [reply pyproject.toml]: done

File: pychromecast/socket_client.py (hunk 2 above)
Comment (geoffoxholm) [reply pychromecast/socket_client.py:495]: > Issue https://github.com/home-assistant-libs/pychromecast/issues/738 talks about the Sirius XM app misbehaving, but your comment in the PR talks about Audible. Could you clarify the relationship between the two apps? Or are they the same?

They are different apps. In the end the issue with Sirius XM was "fixed" (avoided) when the app was redesigned. It seems the Audible app has a similar pathology to the one that was present in the Sirius XM app, but no longer is.

Comment (aryasenna) [reply pychromecast/socket_client.py:495]: Hi @emontnemery,

Any insight from the last logs you requested from @geoffoxholm?

I also have similar problem and setup and would like to contribute so that Audible works when pychromecast is active on the same network. 

I can test the changes fix some logic.

As of now the sleep tricks works, but I understand in your judgement it wont ever be merged  since it's an ugly trick to fix race condition.

What is the better way to handle this? To me this behaves like an app quirk,and should be handled as so.

**Comment by geoffoxholm** [comment]
I showed the logs to Chat GPT, and it says:

Screenshot 2024-10-13 at 7 43 06 AM

**Comment by emontnemery** [comment]
What Chat GPT is suggesting is what I had in mind too; the delay prevents some kind of overlap between outstanding requests.
Unfortunately, that theory is not backed up by the logs AFAICT, meaning it's just hallucinations.

**Comment by emontnemery** [comment]
Since the logs don't point to the delay preventing an overlap in requests, I think the delay is the best we can do.
We should try to make it more targeted though, we now sleep every time we get a status update.

Could you try this change instead of what's in the PR:
```diff
diff --git a/pychromecast/socket_client.py b/pychromecast/socket_client.py
index 1d8eac1..c3d911e 100644
--- a/pychromecast/socket_client.py
+++ b/pychromecast/socket_client.py
@@ -490,6 +490,14 @@ class SocketClient(threading.Thread, CastStatusListener):
         self.session_id = status.session_id

         if new_channel and self.destination_id is not None:
+            if (
+                any(namespace in self._handlers for namespace in self.app_namespaces)
+                and self.destination_id not in self._open_channels
+                and status.app_id == "25456794"
+            ):
+                self.logger.debug("Detected Audible app. Sleeping for 1s")
+                time.sleep(1)
+
             # If any of the namespaces of the new app are supported
             # we will automatically connect to it to receive updates
             for namespace in self.app_namespaces:
```

**Comment by emontnemery** [comment]
@aryasenna @geoffoxholm could you try the version I suggest in https://github.com/home-assistant-libs/pychromecast/pull/886#issuecomment-2414124947 ?

**Comment by geoffoxholm** [comment]
@emontnemery - yes, sorry, i will try it as soon as i can. thank you for the suggestion.

**Comment by aryasenna** [comment]
> @aryasenna @geoffoxholm could you try the version I suggest in [#886 (comment)](https://github.com/home-assistant-libs/pychromecast/pull/886#issuecomment-2414124947) ?

Thanks a lot for the response.

I've tested your proposed code on my HA system.
Your targeted `if` still detects Audible's new cast status just fine:

````
...

 if new_channel and self.destination_id is not None:
            if (
                any(namespace in self._handlers for namespace in self.app_namespaces)
                and self.destination_id not in self._open_channels
                and status.app_id == "25456794"
            ):
                self.logger.warning("Checking emontnemery patch")
                self.logger.warning("Detected Audible app. Sleeping for 1s")
                time.sleep(1)

             # If any of the namespaces of the new app are supported
             # we will automatically connect to it to receive updates
             for namespace in self.app_namespaces:


...
````


````
Logger: pychromecast.socket_client
Source: /usr/local/lib/python3.12/site-packages/pychromecast/socket_client.py:501
First occurred: 7:35:39 PM (2 occurrences)
Last logged: 7:36:04 PM

Checking emontnemery patch

````
````
Logger: pychromecast.socket_client
Source: /usr/local/lib/python3.12/site-packages/pychromecast/socket_client.py:501
First occurred: 7:35:40 PM (2 occurrences)
Last logged: 7:36:05 PM

Detected Audible app. Sleeping for 1s

````

Audible plays fine on the target cast device (tested with Nest Hub and Chromecast Audio)

After checking the patch, I get your hesitancy, it's quite "ugly" to make sleep based on an app ID like this, but it works and Audible is not so obscure app if you see #https://github.com/home-assistant/core/issues/87387#issuecomment-2413198985 🤷🏾‍♂️

**Comment by geoffoxholm** [comment]
@aryasenna - thank you for checking. i have added you to my fork, so if you wish to push to the PR branch, please go ahead.

**Comment by aryasenna** [comment]
@emontnemery I integrated your code to the PR [d418837](https://github.com/home-assistant-libs/pychromecast/pull/886/commits/d418837643e87021a2c0b1f953eccac9977d5bf6)

File: pychromecast/socket_client.py [hunk 3]
```diff
@@ -489,6 +489,16 @@ def new_cast_status(self, status: CastStatus) -> None:
         self.destination_id = status.transport_id
         self.session_id = status.session_id
 
+        # App quirk: when a new cast status is received, give Audible time to load
```
Comment (emontnemery) [inline pychromecast/socket_client.py:492]:
```suggestion
        # App quirk: If the Audible app is running, wait 1s after receiving cast status
        # before connecting to the media player channel. This is a workaround for 
        # https://github.com/home-assistant-libs/pychromecast/issues/738
```

**Comment by emontnemery** [comment]
@aryasenna You didn't try my suggestion, I meant for it the delay to be moved inside the check for new channel.
Maybe reading the diff is not that easy, here's the complete `new_cast_status` method:

```py
    def new_cast_status(self, status: CastStatus) -> None:
        """Called when a new cast status has been received."""
        new_channel = self.destination_id != status.transport_id

        if new_channel and self.destination_id is not None:
            self.disconnect_channel(self.destination_id)

        self.app_namespaces = status.namespaces
        self.destination_id = status.transport_id
        self.session_id = status.session_id

        if new_channel and self.destination_id is not None:
            # App quirk: If the Audible app is running, wait 1s after receiving cast status
            # before connecting to the media player channel. This is a workaround for 
            # https://github.com/home-assistant-libs/pychromecast/issues/738
            if (
                any(namespace in self._handlers for namespace in self.app_namespaces)
                    and self.destination_id not in self._open_channels
                    and status.app_id == "25456794"
                ):

                self.logger.debug("Detected Audible connection. Sleeping for 1s")
                time.sleep(1)

            # If any of the namespaces of the new app are supported
            # we will automatically connect to it to receive updates
            for namespace in self.app_namespaces:
                if namespace in self._handlers:
                    self._ensure_channel_connected(self.destination_id)
                    for handler in set(self._handlers[namespace]):
                        handler.channel_connected()
```

**Comment by emontnemery** [comment]
Also, instead of the magic number, please add the app constant:
```diff
diff --git a/pychromecast/config.py b/pychromecast/config.py
index 9c9864e..ac5a6ac 100644
--- a/pychromecast/config.py
+++ b/pychromecast/config.py
@@ -22,6 +22,7 @@ APP_BBCIPLAYER = "5E81F6DB"
 APP_SHAKA = "07AEE832"
 APP_NRKTV = "3AEDF8D1"
 APP_NRKRADIO = "A49874B1"
+APP_AUDIBLE = "25456794"
```

**Comment by aryasenna** [comment]
@geoffoxholm Sorry for the confusion. I basically updated the PR branch with the wrong code in d418837643e87021a2c0b1f953eccac9977d5bf6, 

but when I did the test, the sleep() was definitely under `if new_channel and self.destination_id is not None:`

Just double-checked the current code running on my VM:
image

I can confirm with the current code Audible is working fine.

---

I fixed it on e05baea98f57da6ebb49fd28bd3ec471e54547d2, and it includes the use of app constant, as you requested.

**Comment by geoffoxholm** [comment]
Thanks @aryasenna for following through on this. I have reviewed your changes locally, and have tested in my own environment and everything looks good to me.

as an _extremely minor_ nit, the blank line after the conditional is unusual for this codebase.

**Comment by aryasenna** [comment]
OK @geoffoxholm, I've also removed the extra line break.

 I hope @emontnemery finds this PR acceptable now,

**Review by emontnemery — APPROVED** [review]
Thanks a lot @geoffoxholm and @aryasenna 👍

**Comment by emontnemery** [comment]
Home assistant core PR: <https://github.com/home-assistant/core/pull/129251>
Will land in Home Assistant core 2024.11.0
